packages/admin/src/stack/Stack.tsx

```diff
@@ -55,211 +50,142 @@ const sortByParentId = <TSortNode extends ISortNode>(nodes: TSortNode[]) => {
         }
-    }
-
-    private goAllBack() {
-        this.history.push(this.state.breadcrumbs[0].url);
-    }
+    };
 
-    private addBreadcrumb(id: string, parentId: string, url: string, title: React.ReactNode, invisible: boolean) {
-        this.setState((state) => {
-            const breadcrumbs = [
-                ...state.breadcrumbs,
-                {
-                    id,
-                    parentId,
-                    url,
-                    title,
-                    invisible,
-                },
-            ];
+    const goAllBack = () => {
+        historyRef.current?.push(breadCrumbs[0].url);
+    };
 
-            return {
-                breadcrumbs,
-            };
-        });
-    }
+    const addBreadcrumb = (id: string, parentId: string, url: string, title: React.ReactNode, invisible: boolean) => {
+        setBreadCrumbs([
+            ...breadCrumbs,
+            {
+                id,
+                parentId,
+                url,
+                title,
+                invisible,
+            },
+        ]);
+    };
 
-    private updateBreadcrumb(id: string, parentId: string, url: string, title: React.ReactNode, invisible: boolean) {
-        this.setState((state) => {
-            const breadcrumbs = state.breadcrumbs.map((crumb) => {
+    const updateBreadcrumb = (id: string, parentId: string, url: string, title: React.ReactNode, invisible: boolean) => {
+        setBreadCrumbs(
+            breadCrumbs.map((crumb) => {
                 return crumb.id === id ? { id, parentId, url, title, invisible } : crumb;
-            });
-            return {
-                breadcrumbs,
-            };
-        });
-    }
+            }),
+        );
+    };
 
-    private removeBreadcrumb(id: string) {
-        this.setState((state) => {
-            const breadcrumbs = state.breadcrumbs.filter((crumb) => {
-                return crumb.id !== id;
-            });
-            return {
-                breadcrumbs,
-            };
+    const removeBreadcrumb = (id: string) => {
+        const index = breadCrumbs.findIndex((crumb) => {
+            return crumb.id === id;
         });
-    }
+        const filteredBreadcrumbs = breadCrumbs.splice(index, 1);
+        setBreadCrumbs(filteredBreadcrumbs);
+    };
 
-    private addSwitchMeta = (id: string, options: { parentId: string; activePage: string; isInitialPageActive: boolean }) => {
-        this.setState((state) => {
-            const switches = [...state.switches];
-            const index = switches.findIndex((i) => i.id === id);
-            if (index === -1) {
-                switches.push({ id, ...options });
-            } else {
-                switches[index] = { id, ...options };
-            }
-            return {
-                switches,
-            };
-        });
+    const addSwitchMeta = (id: string, options: { parentId: string; activePage: string; isInitialPageActive: boolean }) => {
```

Comment:
All those methods are refactored to functions. Methods are not recreated during renders, but functions are.

To make the component behave as close as the original class-component-version does (that is what you most likely want when you refactor), you could memoize all functions with `useCallback`.
All those functions have zero dependencies `useCallback(()=> ..., [ ]) // deps array is empty`, which makes them perfect candidates to memoize.

To access the previous state of switches/breadcrumbs you can pass a function to setState... `setSwitches( previousSwitches => newSwitches )`, so addSwitchMeta is no closure anymore (closures are harder to memoize).